fmriprep/interfaces/cifti.py

```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+"""
+Handling connectivity
+~~~~~~~~~~~~~~~~~~~~~
+Combines FreeSurfer surfaces with subcortical volumes
+
+"""
+import os
+from glob import glob
+
+import nibabel as nb
+from nibabel import cifti2 as ci
+import numpy as np
+
+from niworkflows.nipype.interfaces.base import (
+    BaseInterfaceInputSpec, TraitedSpec, File, traits, isdefined,
+    SimpleInterface, Directory
+)
+
+from niworkflows.nipype.utils.filemanip import split_filename
+
+# CITFI structures with corresponding FS labels
+CIFTI_STRUCT_WITH_LABELS = {
+    # SURFACES
+    'CIFTI_STRUCTURE_CORTEX_LEFT': [],
+    'CIFTI_STRUCTURE_CORTEX_RIGHT': [],
+    # 'CIFTI_STRUCTURE_CORTEX': [],
+    # 'CIFTI_STRUCTURE_CEREBRAL_WHITE_MATTER_LEFT': [],
+    # 'CIFTI_STRUCTURE_CEREBRAL_WHITE_MATTER_RIGHT': [],
+
+    # SUBCORTICAL
+    'CIFTI_STRUCTURE_ACCUMBENS_LEFT': [26],
+    'CIFTI_STRUCTURE_ACCUMBENS_RIGHT': [58],
+    # 'CIFTI_STRUCTURE_ALL_WHITE_MATTER': [],
+    # 'CIFTI_STRUCTURE_ALL_GREY_MATTER': [],
+    'CIFTI_STRUCTURE_AMYGDALA_LEFT': [18],
+    'CIFTI_STRUCTURE_AMYGDALA_RIGHT': [54],
+    'CIFTI_STRUCTURE_BRAIN_STEM': [16],
+    'CIFTI_STRUCTURE_CAUDATE_LEFT': [11],
+    'CIFTI_STRUCTURE_CAUDATE_RIGHT': [50],
+    'CIFTI_STRUCTURE_CEREBELLAR_WHITE_MATTER_LEFT': [7],
+    'CIFTI_STRUCTURE_CEREBELLAR_WHITE_MATTER_RIGHT': [46],
+    # 'CIFTI_STRUCTURE_CEREBELLUM': [],
+    'CIFTI_STRUCTURE_CEREBELLUM_LEFT': [6],
+    'CIFTI_STRUCTURE_CEREBELLUM_RIGHT': [45],
+    'CIFTI_STRUCTURE_DIENCEPHALON_VENTRAL_LEFT': [28],
+    'CIFTI_STRUCTURE_DIENCEPHALON_VENTRAL_RIGHT': [60],
+    'CIFTI_STRUCTURE_HIPPOCAMPUS_LEFT': [17],
+    'CIFTI_STRUCTURE_HIPPOCAMPUS_RIGHT': [53],
+    # 'CIFTI_STRUCTURE_OTHER': [],
+    # 'CIFTI_STRUCTURE_OTHER_GREY_MATTER': [],
+    # 'CIFTI_STRUCTURE_OTHER_WHITE_MATTER': [],
+    'CIFTI_STRUCTURE_PALLIDUM_LEFT': [13],
+    'CIFTI_STRUCTURE_PALLIDUM_RIGHT': [52],
+    'CIFTI_STRUCTURE_PUTAMEN_LEFT': [12],
+    'CIFTI_STRUCTURE_PUTAMEN_RIGHT': [51],
+    'CIFTI_STRUCTURE_THALAMUS_LEFT': [10],
+    'CIFTI_STRUCTURE_THALAMUS_RIGHT': [49],
+}
+
+
+class GenerateCiftiInputSpec(BaseInterfaceInputSpec):
+    bold_file = File(mandatory=True, exists=True, desc="input BOLD file")
+    volume_target = traits.Enum("MNI152NLin2009cAsym", mandatory=True, usedefault=True,
+                                desc="CIFTI volumetric output space")
+    surface_target = traits.Enum("fsaverage5", "fsaverage6", "fsaverage", mandatory=True,
+                                 usedefault=True, desc="CIFTI surface target space")
+    subjects_dir = Directory(mandatory=True, desc="FreeSurfer SUBJECTS_DIR")
+    TR = traits.Float(mandatory=True, desc="repetition time")
+    gifti_files = traits.List(File(exists=True), mandatory=True,
+                              desc="list of surface geometry files (length 2 with order [L,R])")
```

Comment:
@effigies  - does the gifti file contain any info about the surface target, in which case surface_target should be removed and just derived from the gifti file.

Reply:
So these are functional time series, and it does not appear so.

```Shell
% mris_info /data/out/ds000114/derivatives/fmriprep/sub-07/ses-test/func/sub-07_ses-test_task-fingerfootlips_bold_space-fsaverage5.L.func.gii
gifti_image struct
    version    = 1.0
    numDA      = 184
gim->meta nvpairs struct, len = 4 :
    nvpair: 'AnatomicalStructurePrimary' = 'CortexLeft'
    nvpair: 'UserName' = 'root'
    nvpair: 'Date' = 'Mon Oct 16 18:34:03 2017'
    nvpair: 'gifticlib-version' = 'gifti library version 1.09, 28 June, 2010'

gim->labeltable giiLabelTable struct, len = 0 :
--------------------------------------------------
gim->darray[0] giiDataArray struct
    intent   2001 = NIFTI_INTENT_TIME_SERIES
    datatype   16 = NIFTI_TYPE_FLOAT32
    ind_ord     1 = RowMajorOrder
    num_dim       = 1
    dims          = 10242, 0, 0, 0, 0, 0
    encoding    3 = GZipBase64Binary
    endian      2 = LittleEndian
    ext_fname     = 
    ext_offset    = 0
darray->meta nvpairs struct, len = 1 :
    nvpair: 'TimeStep' = '2500.000000'

darray->coordsys giiCoordSystem struct
    dataspace  = NIFTI_XFORM_UNKNOWN
    xformspace = NIFTI_XFORM_UNKNOWN
    xform[0] :  1.000000  0.000000  0.000000  0.000000
    xform[1] :  0.000000  1.000000  0.000000  0.000000
    xform[2] :  0.000000  0.000000  1.000000  0.000000
    xform[3] :  0.000000  0.000000  0.000000  1.000000
    data       = <set>
    nvals      = 10242
    nbyper     = 4
    numCS      = 1
darray->ex_atrs nvpairs struct, len = 0 :
--------------------------------------------------
--------------------------------------------------
gim->darray[1] giiDataArray struct
    intent   2001 = NIFTI_INTENT_TIME_SERIES
    datatype   16 = NIFTI_TYPE_FLOAT32
    ind_ord     1 = RowMajorOrder
    num_dim       = 1
    dims          = 10242, 0, 0, 0, 0, 0
    encoding    3 = GZipBase64Binary
    endian      2 = LittleEndian
    ext_fname     = 
    ext_offset    = 0
darray->meta nvpairs struct, len = 1 :
    nvpair: 'TimeStep' = '2500.000000'

darray->coordsys giiCoordSystem struct
    dataspace  = NIFTI_XFORM_UNKNOWN
    xformspace = NIFTI_XFORM_UNKNOWN
    xform[0] :  1.000000  0.000000  0.000000  0.000000
    xform[1] :  0.000000  1.000000  0.000000  0.000000
    xform[2] :  0.000000  0.000000  1.000000  0.000000
    xform[3] :  0.000000  0.000000  0.000000  1.000000
    data       = <set>
    nvals      = 10242
    nbyper     = 4
    numCS      = 1
darray->ex_atrs nvpairs struct, len = 0 :
--------------------------------------------------
[...]
--------------------------------------------------
gim->darray[183] giiDataArray struct
    intent   2001 = NIFTI_INTENT_TIME_SERIES
    datatype   16 = NIFTI_TYPE_FLOAT32
    ind_ord     1 = RowMajorOrder
    num_dim       = 1
    dims          = 10242, 0, 0, 0, 0, 0
    encoding    3 = GZipBase64Binary
    endian      2 = LittleEndian
    ext_fname     = 
    ext_offset    = 0
darray->meta nvpairs struct, len = 1 :
    nvpair: 'TimeStep' = '2500.000000'

darray->coordsys giiCoordSystem struct
    dataspace  = NIFTI_XFORM_UNKNOWN
    xformspace = NIFTI_XFORM_UNKNOWN
    xform[0] :  1.000000  0.000000  0.000000  0.000000
    xform[1] :  0.000000  1.000000  0.000000  0.000000
    xform[2] :  0.000000  0.000000  1.000000  0.000000
    xform[3] :  0.000000  0.000000  0.000000  1.000000
    data       = <set>
    nvals      = 10242
    nbyper     = 4
    numCS      = 1
darray->ex_atrs nvpairs struct, len = 0 :
--------------------------------------------------
gifti_image struct
    swapped    = 0
    compressed = 1
 -- darray totals: 7 MB
gim->ex_atrs nvpairs struct, len = 0 :
==================================================
```